File: Cyberbullying-detection/08_dashboard/js/csv_upload.js
/**
 * CSV Upload Module for Cyberbullying Detection Dashboard
 * Handles file upload, validation, preview, and analysis
 */

const CSVUpload = (function() {
    'use strict';

    // State
    let state = {
        selectedFile: null,
        fileData: null,
        analysisResults: null,
        isProcessing: false
    };

    // DOM Elements cache
    let elements = {};

    /**
     * Initialize CSV Upload module
     */
    function init() {
        console.log('Initializing CSV Upload module...');
        cacheElements();
        setupEventListeners();
    }

    /**
     * Cache DOM elements
     */
    function cacheElements() {
        elements = {
            // Buttons
            uploadBtn: document.getElementById('upload-csv-btn'),
            analyzeBtn: document.getElementById('analyze-btn'),
            cancelBtn: document.getElementById('cancel-upload'),
            removeFileBtn: document.getElementById('remove-file'),
            retryBtn: document.getElementById('retry-btn'),
            loadResultsBtn: document.getElementById('load-results-btn'),
            closeModalBtn: document.getElementById('close-csv-modal'),
            
            // Modal
            modal: document.getElementById('csv-upload-modal'),
            
            // Sections
            uploadSection: document.getElementById('upload-section'),
            previewSection: document.getElementById('preview-section'),
            processingSection: document.getElementById('processing-section'),
            resultsSection: document.getElementById('results-section'),
            errorSection: document.getElementById('error-section'),
            
            // Upload elements
            dropzone: document.getElementById('dropzone'),
            fileInput: document.getElementById('csv-file-input'),
            
            // File info
            fileName: document.getElementById('selected-file-name'),
            fileSize: document.getElementById('selected-file-size'),
            
            // Selects
            modelSelect: document.getElementById('model-select'),
            textColumnSelect: document.getElementById('text-column-select'),
            
            // Preview table
            previewThead: document.getElementById('preview-thead'),
            previewTbody: document.getElementById('preview-tbody'),
            
            // Progress
            progressFill: document.getElementById('progress-fill'),
            progressText: document.getElementById('progress-text'),
            
            // Results
            resultTotal: document.getElementById('result-total'),
            resultCyberbullying: document.getElementById('result-cyberbullying'),
            resultSafe: document.getElementById('result-safe'),
            resultTime: document.getElementById('result-time'),
            
            // Severity bars
            barCritical: document.getElementById('bar-critical'),
            barHigh: document.getElementById('bar-high'),
            barMedium: document.getElementById('bar-medium'),
            barLow: document.getElementById('bar-low'),
            countCritical: document.getElementById('count-critical'),
            countHigh: document.getElementById('count-high'),
            countMedium: document.getElementById('count-medium'),
            countLow: document.getElementById('count-low'),
            
            // Error
            errorMessage: document.getElementById('error-message')
        };
    }

    /**
     * Setup event listeners
     */
    function setupEventListeners() {
        // Open modal
        if (elements.uploadBtn) {
            elements.uploadBtn.addEventListener('click', openModal);
        }

        // Close modal
        if (elements.closeModalBtn) {
            elements.closeModalBtn.addEventListener('click', closeModal);
        }
        if (elements.cancelBtn) {
            elements.cancelBtn.addEventListener('click', closeModal);
        }
        if (elements.modal) {
            elements.modal.addEventListener('click', (e) => {
                if (e.target === elements.modal) closeModal();
            });
        }

        // Dropzone
        if (elements.dropzone) {
            elements.dropzone.addEventListener('click', () => elements.fileInput?.click());
            elements.dropzone.addEventListener('dragover', handleDragOver);
            elements.dropzone.addEventListener('dragleave', handleDragLeave);
            elements.dropzone.addEventListener('drop', handleDrop);
        }

        // File input
        if (elements.fileInput) {
            elements.fileInput.addEventListener('change', handleFileSelect);
        }

        // Remove file
        if (elements.removeFileBtn) {
            elements.removeFileBtn.addEventListener('click', removeFile);
        }

        // Analyze
        if (elements.analyzeBtn) {
            elements.analyzeBtn.addEventListener('click', analyzeCSV);
        }

        // Retry
        if (elements.retryBtn) {
            elements.retryBtn.addEventListener('click', resetToUpload);
        }

        // Load results
        if (elements.loadResultsBtn) {
            elements.loadResultsBtn.addEventListener('click', loadResultsToDashboard);
        }

        // Keyboard
        document.addEventListener('keydown', (e) => {
            if (e.key === 'Escape' && !elements.modal?.classList.contains('hidden')) {
                closeModal();
            }
        });
    }

    /**
     * Open upload modal
     */
    function openModal() {
        elements.modal?.classList.remove('hidden');
        resetToUpload();
        document.body.style.overflow = 'hidden';
    }

    /**
     * Close upload modal
     */
    function closeModal() {
        elements.modal?.classList.add('hidden');
        document.body.style.overflow = '';
        if (!state.isProcessing) {
            resetState();
        }
    }

    /**
     * Reset to upload section
     */
    function resetToUpload() {
        hideAllSections();
        elements.uploadSection?.classList.remove('hidden');
        elements.analyzeBtn?.classList.remove('hidden');
        elements.loadResultsBtn?.classList.add('hidden');
        if (elements.analyzeBtn) elements.analyzeBtn.disabled = true;
    }

    /**
     * Reset state
     */
    function resetState() {
        state = {
            selectedFile: null,
            fileData: null,
            analysisResults: null,
            isProcessing: false
        };
        if (elements.fileInput) elements.fileInput.value = '';
    }

    /**
     * Hide all sections
     */
    function hideAllSections() {
        elements.uploadSection?.classList.add('hidden');
        elements.previewSection?.classList.add('hidden');
        elements.processingSection?.classList.add('hidden');
        elements.resultsSection?.classList.add('hidden');
        elements.errorSection?.classList.add('hidden');
    }

    /**
     * Handle drag over
     */
    function handleDragOver(e) {
        e.preventDefault();
        e.stopPropagation();
        elements.dropzone?.classList.add('dragover');
    }

    /**
     * Handle drag leave
     */
    function handleDragLeave(e) {
        e.preventDefault();
        e.stopPropagation();
        elements.dropzone?.classList.remove('dragover');
    }

    /**
     * Handle drop
     */
    function handleDrop(e) {
        e.preventDefault();
        e.stopPropagation();
        elements.dropzone?.classList.remove('dragover');

        const files = e.dataTransfer?.files;
        if (files && files.length > 0) {
            processFile(files[0]);
        }
    }

    /**
     * Handle file select from input
     */
    function handleFileSelect(e) {
        const files = e.target.files;
        if (files && files.length > 0) {
            processFile(files[0]);
        }
    }

    /**
     * Process selected file
     */
    async function processFile(file) {
        // Validate file type (CSV or TXT)
        const fileName = file.name.toLowerCase();
        if (!fileName.endsWith('.csv') && !fileName.endsWith('.txt')) {
            showError('Invalid file format. Please upload a CSV or TXT file.');
            return;
        }

        // Validate file size (10MB max)
        if (file.size > 10 * 1024 * 1024) {
            showError('File too large. Maximum size is 10MB.');
            return;
        }

        state.selectedFile = file;

        // Update file info
        if (elements.fileName) {
            elements.fileName.textContent = file.name;
        }
        if (elements.fileSize) {
            elements.fileSize.textContent = `(${formatFileSize(file.size)})`;
        }

        // Parse and preview CSV
        try {
            await parseCSV(file);
            showPreview();
        } catch (error) {
            showError(`Failed to parse CSV: ${error.message}`);
        }
    }

    /**
     * Parse CSV or TXT file
     */
    async function parseCSV(file) {
        return new Promise((resolve, reject) => {
            const reader = new FileReader();
            const isTxt = file.name.toLowerCase().endsWith('.txt');
            
            reader.onload = (e) => {
                try {
                    const text = e.target.result;
                    const lines = text.split('\n').filter(line => line.trim());
                    
                    if (lines.length < 1) {
                        reject(new Error('File must have at least one line of data'));
                        return;
                    }

                    let headers, data;

                    if (isTxt) {
                        // TXT file: each line is a message
                        headers = ['line_number', 'message'];
                        data = lines.slice(0, 100).map((line, idx) => ({
                            line_number: idx + 1,
                            message: line.trim()
                        }));
                    } else {
                        // CSV file: parse headers and rows
                        if (lines.length < 2) {
                            reject(new Error('CSV must have at least a header and one data row'));
                            return;
                        }

                        headers = parseCSVLine(lines[0]);
                        data = [];
                        for (let i = 1; i < lines.length && i <= 100; i++) {
                            const values = parseCSVLine(lines[i]);
                            if (values.length === headers.length) {
                                const row = {};
                                headers.forEach((h, idx) => {
                                    row[h] = values[idx];
                                });
                                data.push(row);
                            }
                        }
                    }

                    state.fileData = {
                        headers,
                        data,
                        totalRows: isTxt ? lines.length : lines.length - 1,
                        isTxt
                    };

                    // Populate text column select
                    populateTextColumnSelect(headers);

                    resolve(state.fileData);
                } catch (err) {
                    reject(err);
                }
            };

            reader.onerror = () => reject(new Error('Failed to read file'));
            reader.readAsText(file);
        });
    }

    /**
     * Parse a single CSV line
     */
    function parseCSVLine(line) {
        const result = [];
        let current = '';
        let inQuotes = false;

        for (let i = 0; i < line.length; i++) {
            const char = line[i];
            
            if (char === '"') {
                inQuotes = !inQuotes;
            } else if (char === ',' && !inQuotes) {
                result.push(current.trim());
                current = '';
            } else {
                current += char;
            }
        }
        result.push(current.trim());
        
        return result;
    }

    /**
     * Populate text column select with auto-detection
     */
    function populateTextColumnSelect(headers) {
        if (!elements.textColumnSelect) return;

        // Priority text column names
        const textColumnPriority = ['message', 'text', 'content', 'chat', 'comment', 'msg', 'body', 'post', 'tweet'];
        // Metadata columns to skip
        const metadataColumns = ['name', 'username', 'user', 'sender', 'from', 'author', 'time', 'timestamp', 
                                  'datetime', 'date', 'created_at', 'id', 'chat_id', 'message_id', 'student_id',
                                  'user_id', 'platform', 'source', 'phone', 'email', 'number'];

        elements.textColumnSelect.innerHTML = '<option value="">Auto-detect</option>';
        
        let detectedColumn = null;
        
        headers.forEach(header => {
            const option = document.createElement('option');
            option.value = header;
            option.textContent = header;
            elements.textColumnSelect.appendChild(option);
            
            // Auto-detect best text column
            const headerLower = header.toLowerCase();
            if (!detectedColumn) {
                if (textColumnPriority.some(tc => headerLower.includes(tc))) {
                    detectedColumn = header;
                }
            }
        });

        // If detected, pre-select it
        if (detectedColumn) {
            elements.textColumnSelect.value = detectedColumn;
            console.log(`Auto-detected text column: ${detectedColumn}`);
        }
    }

    /**
     * Show preview section
     */
    function showPreview() {
        hideAllSections();
        elements.previewSection?.classList.remove('hidden');
        
        if (elements.analyzeBtn) {
            elements.analyzeBtn.disabled = false;
        }

        renderPreviewTable();
    }

    /**
     * Render preview table (highlighting the message column)
     */
    function renderPreviewTable() {
        if (!state.fileData || !elements.previewThead || !elements.previewTbody) return;

        const { headers, data, isTxt } = state.fileData;
        
        // Get the selected/detected text column
        const selectedCol = elements.textColumnSelect?.value || '';
        const textColumnPriority = ['message', 'text', 'content', 'chat', 'comment', 'msg', 'body'];
        
        // Determine which column is the message column
        let messageCol = selectedCol;
        if (!messageCol) {
            for (const col of textColumnPriority) {
                if (headers.map(h => h.toLowerCase()).includes(col)) {
                    messageCol = headers[headers.map(h => h.toLowerCase()).indexOf(col)];
                    break;
                }
            }
        }

        // Render headers (highlight message column)
        elements.previewThead.innerHTML = `
            <tr>
                ${headers.map(h => `<th class="${h === messageCol ? 'highlight-col' : ''}">${escapeHtml(h)}${h === messageCol ? ' ✓' : ''}</th>`).join('')}
            </tr>
        `;

        // Render first 5 rows (highlight message column)
        const previewData = data.slice(0, 5);
        elements.previewTbody.innerHTML = previewData.map(row => `
            <tr>
                ${headers.map(h => `<td class="${h === messageCol ? 'highlight-col' : ''}">${escapeHtml(truncate(row[h] || '', 50))}</td>`).join('')}
            </tr>
        `).join('');
    }

    /**
     * Remove selected file
     */
    function removeFile() {
        resetState();
        resetToUpload();
    }

    /**
     * Analyze CSV
     * FIX: Added proper logging and ensures ALL messages get classified
     */
    async function analyzeCSV() {
        if (!state.selectedFile || state.isProcessing) return;

        console.log('🔍 Starting CSV analysis...');
        state.isProcessing = true;
        
        hideAllSections();
        elements.processingSection?.classList.remove('hidden');
        elements.analyzeBtn?.classList.add('hidden');

        // Progress tracking
        let progress = 0;
        const progressInterval = setInterval(() => {
            progress += Math.random() * 5;
            if (progress > 85) progress = 85;
            updateProgress(progress, 'Analyzing messages...');
        }, 300);

        try {
            const formData = new FormData();
            formData.append('file', state.selectedFile);
            formData.append('model_type', elements.modelSelect?.value || 'bert');
            
            const textColumn = elements.textColumnSelect?.value;
            if (textColumn) {
                formData.append('text_column', textColumn);
            }

            console.log('📤 Uploading file to backend for analysis...');
            const response = await fetch(`${APIClient.getBaseUrl()}/upload/csv`, {
                method: 'POST',
                body: formData
            });

            clearInterval(progressInterval);

            if (!response.ok) {
                const error = await response.json();
                throw new Error(error.detail || 'Analysis failed');
            }

            const results = await response.json();
            
            // Verify ALL messages were classified
            console.log(`✅ Analysis complete. Checking results...`);
            console.log(`   Total rows: ${results.analyzed_rows}`);
            console.log(`   Results returned: ${results.results?.length || 0}`);
            
            // FIX: Ensure all messages have classification results
            if (results.results && results.results.length > 0) {
                const classifiedCount = results.results.filter(r => r.prediction !== undefined).length;
                const unclassifiedCount = results.results.length - classifiedCount;
                
                console.log(`   Classified: ${classifiedCount}`);
                if (unclassifiedCount > 0) {
                    console.warn(`   ⚠️ Unclassified: ${unclassifiedCount} - These may need re-processing`);
                }
            }
            
            state.analysisResults = results;

            updateProgress(100, 'Complete!');
            console.log('✅ CSV analysis finished successfully');
            
            setTimeout(() => {
                showResults(results);
            }, 500);

        } catch (error) {
            clearInterval(progressInterval);
            console.error('Analysis error:', error);
            showError(error.message || 'Failed to analyze CSV');
        } finally {
            state.isProcessing = false;
        }
    }

    /**
     * Update progress bar
     */
    function updateProgress(percent, text) {
        if (elements.progressFill) {
            elements.progressFill.style.width = `${percent}%`;
        }
        if (elements.progressText) {
            elements.progressText.textContent = text;
        }
    }

    /**
     * Show results
     */
    function showResults(results) {
        hideAllSections();
        elements.resultsSection?.classList.remove('hidden');
        elements.loadResultsBtn?.classList.remove('hidden');
        
        // Update summary stats
        if (elements.resultTotal) {
            elements.resultTotal.textContent = results.analyzed_rows;
        }
        if (elements.resultCyberbullying) {
            elements.resultCyberbullying.textContent = results.cyberbullying_count;
        }
        if (elements.resultSafe) {
            elements.resultSafe.textContent = results.not_cyberbullying_count;
        }
        if (elements.resultTime) {
            elements.resultTime.textContent = `${Math.round(results.processing_time_ms)}ms`;
        }

        // Update severity bars
        const severity = results.severity_distribution || {};
        const total = results.analyzed_rows || 1;

        updateSeverityBar('critical', severity.critical || 0, total);
        updateSeverityBar('high', severity.high || 0, total);
        updateSeverityBar('medium', severity.medium || 0, total);
        updateSeverityBar('low', severity.low || 0, total);
    }

    /**
     * Update severity bar
     */
    function updateSeverityBar(level, count, total) {
        const barEl = elements[`bar${capitalize(level)}`];
        const countEl = elements[`count${capitalize(level)}`];
        
        const percent = (count / total) * 100;
        
        if (barEl) {
            barEl.style.width = `${percent}%`;
        }
        if (countEl) {
            countEl.textContent = count;
        }
    }

    /**
     * Load results to dashboard
     * FIX: Now ensures all messages have proper classification data
     */
    function loadResultsToDashboard() {
        if (!state.analysisResults) {
            console.error('❌ No analysis results to load');
            return;
        }

        console.log('📊 Loading results to dashboard...');
        const results = state.analysisResults;

        // Transform results for dashboard - ensure ALL messages are processed
        const messages = results.results.map((r, idx) => {
            // FIX: Ensure each message has proper classification fields
            const message = {
                id: r.row_id || idx + 1,
                date: new Date().toISOString().split('T')[0],
                student_id: r.student_id || `User_${r.row_id || idx + 1}`,
                message: r.original_text || r.text || '',
                platform: r.platform || 'csv_upload',
                severity: r.severity || (r.is_cyberbullying ? 'high' : 'low'),
                score: r.prediction || (r.is_cyberbullying ? 'Cyberbullying' : 'Safe'),
                confidence: typeof r.confidence === 'number' 
                    ? `${Math.round(r.confidence * 100)}%` 
                    : (r.confidence || 'N/A'),
                prediction: r.prediction,
                model: r.model_used || 'bert',
                is_cyberbullying: r.is_cyberbullying,
                raw_confidence: r.confidence
            };
            
            console.log(`  ✓ Message ${idx + 1}: ${message.severity} (${message.confidence})`);
            return message;
        });

        console.log(`📋 Processed ${messages.length} messages for dashboard`);
        
        // Verify all messages have classifications
        const classified = messages.filter(m => m.severity !== 'unknown' && m.score !== 'N/A');
        const unclassified = messages.filter(m => m.severity === 'unknown' || m.score === 'N/A');
        
        if (unclassified.length > 0) {
            console.warn(`⚠️ ${unclassified.length} messages may not be fully classified`);
        }
        console.log(`✅ ${classified.length}/${messages.length} messages fully classified`);

        // Update dashboard with new data
        if (window.Dashboard) {
            Dashboard.loadCSVResults(messages, results);
        }

        // Close modal
        closeModal();

        // Show notification
        showNotification(`✅ Loaded ${messages.length} analyzed messages to dashboard`);
    }

    /**
     * Show error
     */
    function showError(message) {
        hideAllSections();
        elements.errorSection?.classList.remove('hidden');
        
        if (elements.errorMessage) {
            elements.errorMessage.textContent = message;
        }
        
        elements.analyzeBtn?.classList.remove('hidden');
        if (elements.analyzeBtn) elements.analyzeBtn.disabled = true;
    }

    /**
     * Show notification
     */
    function showNotification(message) {
        // Create toast notification
        const toast = document.createElement('div');
        toast.className = 'toast-notification';
        toast.textContent = message;
        document.body.appendChild(toast);

        setTimeout(() => {
            toast.classList.add('show');
        }, 100);

        setTimeout(() => {
            toast.classList.remove('show');
            setTimeout(() => toast.remove(), 300);
        }, 3000);
    }

    // Utility functions
    function formatFileSize(bytes) {
        if (bytes < 1024) return bytes + ' B';
        if (bytes < 1024 * 1024) return (bytes / 1024).toFixed(1) + ' KB';
        return (bytes / (1024 * 1024)).toFixed(1) + ' MB';
    }

    function escapeHtml(text) {
        const div = document.createElement('div');
        div.textContent = text;
        return div.innerHTML;
    }

    function truncate(str, len) {
        if (!str) return '';
        return str.length > len ? str.substring(0, len) + '...' : str;
    }

    function capitalize(str) {
        return str.charAt(0).toUpperCase() + str.slice(1);
    }

    // Public API
    return {
        init,
        openModal,
        closeModal,
        getResults: () => state.analysisResults
    };
})();

// Initialize when DOM is ready
document.addEventListener('DOMContentLoaded', CSVUpload.init);

// Export for other modules
window.CSVUpload = CSVUpload;

File: Cyberbullying-detection/06_api/routes/csv_upload.py
"""
CSV Upload endpoint for chat analysis.
POST /upload/csv
"""

import io
import csv
import uuid
import time
import logging
from typing import List, Optional
from datetime import datetime
from fastapi import APIRouter, HTTPException, UploadFile, File, Form, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

try:
    from ..models_loader import get_detector
    from ..schemas import PredictionResponse
except ImportError:
    from models_loader import get_detector
    from schemas import PredictionResponse

router = APIRouter()
logger = logging.getLogger(__name__)

# Rate limiting state (simple in-memory)
_rate_limit_store = {}
RATE_LIMIT_WINDOW = 60  # seconds
RATE_LIMIT_MAX_REQUESTS = 10  # max uploads per window per IP


class CSVAnalysisResult(BaseModel):
    """Result for a single analyzed message."""
    row_id: int
    original_text: str
    prediction: str
    confidence: float
    is_cyberbullying: bool
    severity: str
    model_used: str
    probabilities: Optional[dict] = None


class CSVUploadResponse(BaseModel):
    """Response for CSV upload and analysis."""
    success: bool
    message: str
    upload_id: str
    total_rows: int
    analyzed_rows: int
    cyberbullying_count: int
    not_cyberbullying_count: int
    severity_distribution: dict
    processing_time_ms: float
    results: List[CSVAnalysisResult]
    preview: List[dict] = Field(default_factory=list)


def check_rate_limit(ip_address: str) -> bool:
    """Check if request is within rate limit."""
    now = time.time()
    
    # Clean old entries
    _rate_limit_store[ip_address] = [
        t for t in _rate_limit_store.get(ip_address, [])
        if now - t < RATE_LIMIT_WINDOW
    ]
    
    # Check limit
    if len(_rate_limit_store.get(ip_address, [])) >= RATE_LIMIT_MAX_REQUESTS:
        return False
    
    # Add current request
    if ip_address not in _rate_limit_store:
        _rate_limit_store[ip_address] = []
    _rate_limit_store[ip_address].append(now)
    
    return True


def determine_severity(confidence: float, is_cyberbullying: bool) -> str:
    """Determine severity level based on confidence and prediction."""
    if not is_cyberbullying:
        return "low"
    
    if confidence >= 0.9:
        return "critical"
    elif confidence >= 0.75:
        return "high"
    elif confidence >= 0.5:
        return "medium"
    else:
        return "low"


# Columns to EXCLUDE from text detection (metadata columns)
METADATA_COLUMNS = [
    'name', 'username', 'user', 'sender', 'from', 'author',
    'time', 'timestamp', 'datetime', 'date', 'created_at', 'sent_at',
    'id', 'chat_id', 'message_id', 'student_id', 'user_id',
    'platform', 'source', 'channel', 'group',
    'phone', 'number', 'email'
]

import re

# Regex patterns for WhatsApp-style message formats
WHATSAPP_PATTERNS = [
    # Pattern: DD/MM/YYYY, HH:MM am/pm - Name: Message
    re.compile(r'^\d{1,2}/\d{1,2}/\d{2,4},?\s*\d{1,2}:\d{2}(?::\d{2})?\s*(?:am|pm|AM|PM)?\s*-\s*[^:]+:\s*(.+)$'),
    # Pattern: [DD/MM/YYYY, HH:MM:SS] Name: Message
    re.compile(r'^\[\d{1,2}/\d{1,2}/\d{2,4},?\s*\d{1,2}:\d{2}(?::\d{2})?\s*(?:am|pm|AM|PM)?\]\s*[^:]+:\s*(.+)$'),
    # Pattern: YYYY-MM-DD HH:MM - Name: Message
    re.compile(r'^\d{4}-\d{2}-\d{2}\s+\d{1,2}:\d{2}(?::\d{2})?\s*-\s*[^:]+:\s*(.+)$'),
    # Pattern: Name: Message (simple format with name prefix)
    re.compile(r'^[A-Za-z][A-Za-z0-9\s_\-\.]+:\s+(.+)$'),
]


def extract_message_content(text: str) -> str:
    """
    Extract only the actual message content from formatted chat messages.
    Removes date, time, and sender name prefixes (WhatsApp, Telegram, etc.)
    
    Examples:
    - "15/01/2026, 9:49 am - Reshma Begum Akka: Enjoy your festival" -> "Enjoy your festival"
    - "[15/01/2026, 9:49:00 AM] John: Hello" -> "Hello"
    - "User123: Hi there" -> "Hi there"
    """
    if not text:
        return text
    
    text = text.strip()
    
    # Try each pattern
    for pattern in WHATSAPP_PATTERNS:
        match = pattern.match(text)
        if match:
            extracted = match.group(1).strip()
            if extracted:
                return extracted
    
    # If no pattern matched, return original text
    return text

def validate_csv_structure(headers: List[str]) -> tuple:
    """
    Validate CSV structure and find the text/message column.
    Excludes metadata columns like name, time, id, etc.
    Returns (is_valid, text_column_name, error_message)
    """
    # Priority order for text column names
    text_column_names = [
        'message', 'text', 'content', 'chat', 'comment', 
        'msg', 'body', 'post', 'sentence', 'input', 'tweet'
    ]
    
    headers_lower = [h.lower().strip() for h in headers]
    
    # First try to find an exact match for known text columns
    for col_name in text_column_names:
        if col_name in headers_lower:
            idx = headers_lower.index(col_name)
            return True, headers[idx], None
    
    # If no known column found, find the first column that's NOT metadata
    for i, h in enumerate(headers_lower):
        is_metadata = any(meta in h for meta in METADATA_COLUMNS)
        if not is_metadata:
            return True, headers[i], f"Using column '{headers[i]}' as text column (auto-detected)"
    
    # Last resort: use the column with longest average content
    if len(headers) > 0:
        return True, headers[0], f"Using first column '{headers[0]}' as text column"
    
    return False, None, "CSV must have at least one column"


@router.post("/upload/csv", response_model=CSVUploadResponse)
async def upload_csv(
    request: Request,
    file: UploadFile = File(..., description="CSV or TXT file containing chat messages"),
    model_type: str = Form(default="bert", description="Model to use for analysis"),
    text_column: Optional[str] = Form(default=None, description="Name of the column containing text"),
    max_rows: Optional[int] = Form(default=1000, description="Maximum rows to process")
):
    """
    Upload a CSV or TXT file for chat message analysis.
    
    For CSV: Should contain a column with chat messages/text to analyze.
    Common column names detected: message, text, content, chat, comment
    
    For TXT: One message per line.
    
    Returns analysis results with predictions, severity, and confidence scores.
    """
    # Rate limiting
    client_ip = request.client.host if request.client else "unknown"
    if not check_rate_limit(client_ip):
        raise HTTPException(
            status_code=429,
            detail="Rate limit exceeded. Maximum 10 uploads per minute."
        )
    
    # Validate file type
    if not file.filename:
        raise HTTPException(status_code=400, detail="No file uploaded")
    
    filename_lower = file.filename.lower()
    is_txt = filename_lower.endswith('.txt')
    is_csv = filename_lower.endswith('.csv')
    
    if not is_txt and not is_csv:
        raise HTTPException(
            status_code=400, 
            detail="Invalid file format. Please upload a CSV or TXT file."
        )
    
    # Check file size (max 10MB)
    contents = await file.read()
    if len(contents) > 10 * 1024 * 1024:
        raise HTTPException(
            status_code=400,
            detail="File too large. Maximum size is 10MB."
        )
    
    start_time = time.time()
    upload_id = str(uuid.uuid4())
    
    try:
        decoded = contents.decode('utf-8')
        
        if is_txt:
            # Parse TXT file (one message per line)
            # Each line may contain WhatsApp-style format: date, time - name: message
            lines = [line.strip() for line in decoded.split('\n') if line.strip()]
            # Store raw line but message will be extracted during analysis
            rows = [{"message": line} for line in lines]
            headers = ["message"]
            target_column = "message"
            total_rows = len(rows)
        else:
            # Parse CSV
            reader = csv.DictReader(io.StringIO(decoded))
            
            if not reader.fieldnames:
                raise HTTPException(status_code=400, detail="CSV file is empty or has no headers")
            
            headers = list(reader.fieldnames)
            
            # Validate and find text column
            if text_column and text_column in headers:
                target_column = text_column
            else:
                is_valid, target_column, warning = validate_csv_structure(headers)
                if not is_valid:
                    raise HTTPException(status_code=400, detail=warning)
                if warning:
                    logger.info(warning)
            
            # Read all rows
            rows = list(reader)
            total_rows = len(rows)
        
        if total_rows == 0:
            raise HTTPException(status_code=400, detail="CSV file contains no data rows")
        
        # Limit rows
        rows_to_process = rows[:max_rows]
        
        # Get detector
        try:
            detector = get_detector(model_type)
        except Exception as e:
            logger.error(f"Failed to load model: {e}")
            raise HTTPException(
                status_code=500,
                detail=f"Failed to load model '{model_type}'. Available: bert, mbert, indicbert, baseline"
            )
        
        # Analyze messages
        results = []
        cyberbullying_count = 0
        not_cyberbullying_count = 0
        severity_counts = {"low": 0, "medium": 0, "high": 0, "critical": 0}
        
        for idx, row in enumerate(rows_to_process):
            raw_text = row.get(target_column, "").strip()
            
            if not raw_text:
                continue
            
            # Extract only message content (remove date, time, sender name)
            text = extract_message_content(raw_text)
            
            if not text:
                continue
            
            # Log if extraction was applied
            if text != raw_text:
                logger.debug(f"Extracted message: '{raw_text[:50]}...' -> '{text[:50]}...'")
            
            try:
                prediction = detector.predict(text)
                
                is_cb = prediction.get("is_cyberbullying", False)
                confidence = prediction.get("confidence", 0.0)
                severity = determine_severity(confidence, is_cb)
                
                if is_cb:
                    cyberbullying_count += 1
                else:
                    not_cyberbullying_count += 1
                
                severity_counts[severity] += 1
                
                result = CSVAnalysisResult(
                    row_id=idx + 1,
                    original_text=text[:500],  # Truncate long texts
                    prediction=prediction.get("prediction", "Unknown"),
                    confidence=round(confidence, 4),
                    is_cyberbullying=is_cb,
                    severity=severity,
                    model_used=model_type,
                    probabilities=prediction.get("probabilities")
                )
                results.append(result)
                
            except Exception as e:
                logger.warning(f"Failed to analyze row {idx + 1}: {e}")
                continue
        
        # Generate preview (first 5 rows with all columns)
        preview = []
        for row in rows[:5]:
            preview_row = {k: str(v)[:100] for k, v in row.items()}
            preview.append(preview_row)
        
        processing_time = (time.time() - start_time) * 1000
        
        return CSVUploadResponse(
            success=True,
            message=f"Successfully analyzed {len(results)} messages",
            upload_id=upload_id,
            total_rows=total_rows,
            analyzed_rows=len(results),
            cyberbullying_count=cyberbullying_count,
            not_cyberbullying_count=not_cyberbullying_count,
            severity_distribution=severity_counts,
            processing_time_ms=round(processing_time, 2),
            results=results,
            preview=preview
        )
        
    except HTTPException:
        raise
    except UnicodeDecodeError:
        raise HTTPException(
            status_code=400,
            detail="Invalid file encoding. Please use UTF-8 encoded CSV."
        )
    except csv.Error as e:
        raise HTTPException(
            status_code=400,
            detail=f"CSV parsing error: {str(e)}"
        )
    except Exception as e:
        logger.error(f"CSV upload error: {e}")
        raise HTTPException(
            status_code=500,
            detail=f"Internal error processing CSV: {str(e)}"
        )


@router.get("/upload/validate")
async def validate_csv_preview(
    request: Request,
    file: UploadFile = File(..., description="CSV file to validate")
):
    """
    Validate a CSV file without processing.
    Returns headers, row count, and detected text column.
    """
    # Rate limiting
    client_ip = request.client.host if request.client else "unknown"
    if not check_rate_limit(client_ip):
        raise HTTPException(status_code=429, detail="Rate limit exceeded")
    
    if not file.filename or not file.filename.lower().endswith('.csv'):
        raise HTTPException(status_code=400, detail="Invalid file format")
    
    try:
        contents = await file.read()
        decoded = contents.decode('utf-8')
        reader = csv.DictReader(io.StringIO(decoded))
        
        headers = list(reader.fieldnames) if reader.fieldnames else []
        rows = list(reader)
        
        is_valid, text_column, message = validate_csv_structure(headers)
        
        # Preview first 5 rows
        preview = []
        for row in rows[:5]:
            preview.append({k: str(v)[:100] for k, v in row.items()})
        
        return {
            "valid": is_valid,
            "headers": headers,
            "detected_text_column": text_column,
            "total_rows": len(rows),
            "preview": preview,
            "message": message
        }
        
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))
